File: c++/algorithms/two-pointers/two-pointers.cpp
// Dois índices distintos com soma alvo, vetor crescente.
// Entrada: N, N valores e alvo. Ex.: 4 1 3 5 8 9 -> 0 3; ausente -> -1.
// Se a soma é pequena, avance esquerda; se grande, recue direita.
// Cada ponta anda no máximo N vezes: O(N), espaço auxiliar O(1).
// Negativos são permitidos. Não ordene se precisa preservar índices originais.
// Valores e soma devem caber em long long. Não conta todos os pares.
#include <iostream>
#include <utility>
#include <vector>
std::pair<int,int> pair_sum(const std::vector<long long>& v, long long target) {
    int left = 0, right = int(v.size()) - 1;
    while (left < right) {
        long long total = v[left] + v[right];
        if (total == target) return {left, right};
        if (total < target) ++left;
        else --right;
    }
    return {-1, -1};
}
int main() {
    int n; std::cin >> n;
    std::vector<long long> v(n);
    for (auto& x : v) std::cin >> x;
    long long target; std::cin >> target;
    auto [i,j] = pair_sum(v, target);
    if (i == -1) std::cout << -1 << '\n';
    else std::cout << i << ' ' << j << '\n';
}

File: python/algorithms/two_pointers.py
"""Encontra dois índices distintos cuja soma é alvo, em lista ORDENADA.
Se soma é pequena, aumentar esquerda pode aumentá-la; se grande, reduzir direita.
Cada ponteiro anda no máximo N vezes: O(N) tempo, O(1) espaço auxiliar.
Também aceita negativos; não altera a lista. Ordenar antes mudaria os índices!
Entrada: N, valores ordenados e alvo. Ex.: 4 1 3 5 8 9. Saída: 0 3.
Saída -1 se não existir. Não resolve contagem de pares nem qualquer janela.
"""
import sys


def pair_sum(values, target):
    left, right = 0, len(values) - 1
    while left < right:
        total = values[left] + values[right]
        if total == target:
            return left, right
        if total < target:
            left += 1
        else:
            right -= 1
    return None


def main():
    tokens = iter(map(int, sys.stdin.buffer.read().split()))
    n = next(tokens)
    values = [next(tokens) for _ in range(n)]
    answer = pair_sum(values, next(tokens))
    print(*answer) if answer is not None else print(-1)


if __name__ == "__main__":
    main()
